### ISSUE TITLE
The speed of pre- and post- processing based on python backend is very slow!

### ISSUE THREAD

**Issue by WangYizhang01** [triton-inference-server/server#5829]
**Description**
I put all the pre- and post-processing on the triton server, and deploy the yolo model through ensemble. The pre- and post-processing is implemented based on python backend. The deployment is successful, but the speed of the pre- and post- processing is very slow(12s, 23s), even if the request is sent to the pre- and post-processing models separately, the result is the same.I printed the execution time of the execute function in TritonPythonModel, this part is very fast (less than 1s).

**Triton Information**
What version of Triton are you using?
23.03

Are you using the Triton container or did you build it yourself?
the Triton container

**To Reproduce**
Steps to reproduce the behavior.
The following are the pre-processed model.py and config.pbtxt
```python
# model.py
import time

import triton_python_backend_utils as pb_utils
import numpy as np
import cv2


class TritonPythonModel:

    def initialize(self, args):
        pass

    def execute(self, requests):
        t0 = time.time()
        print(f'pre time: {t0}')
        responses = []

        # Every Python backend must iterate over everyone of the requests
        # and create a pb_utils.InferenceResponse for each of them.
        for request in requests:

            def letterbox(img, new_shape=(640, 640), color=(114, 114, 114), auto=True, scaleFill=False, scaleup=True, stride=32):
                # Resize and pad image while meeting stride-multiple constraints
                shape = img.shape[:2]  # current shape [height, width]
                if isinstance(new_shape, int):
                    new_shape = (new_shape, new_shape)

                # Scale ratio (new / old)
                r = min(new_shape[0] / shape[0], new_shape[1] / shape[1])
                if not scaleup:  # only scale down, do not scale up (for better test mAP)
                    r = min(r, 1.0)

                # Compute padding
                ratio = r, r  # width, height ratios
                new_unpad = int(round(shape[1] * r)), int(round(shape[0] * r))
                dw, dh = new_shape[1] - new_unpad[0], new_shape[0] - new_unpad[1]  # wh padding
                if auto:  # minimum rectangle
                    dw, dh = np.mod(dw, stride), np.mod(dh, stride)  # wh padding
                elif scaleFill:  # stretch
                    dw, dh = 0.0, 0.0
                    new_unpad = (new_shape[1], new_shape[0])
                    ratio = new_shape[1] / shape[1], new_shape[0] / shape[0]  # width, height ratios

                dw /= 2  # divide padding into 2 sides
                dh /= 2

                if shape[::-1] != new_unpad:  # resize
                    img = cv2.resize(img, new_unpad, interpolation=cv2.INTER_LINEAR)
                top, bottom = int(round(dh - 0.1)), int(round(dh + 0.1))
                left, right = int(round(dw - 0.1)), int(round(dw + 0.1))
                img = cv2.copyMakeBorder(img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=color)  # add border
                return img, ratio, (dw, dh)

            # Get INPUT0
            batch_img0 = pb_utils.get_input_tensor_by_name(request, "canopycoverage_preprocessing_input")
            batch_img0 = batch_img0.as_numpy()
            res = []
            for i in range(batch_img0.shape[0]):
                img = batch_img0[i]
                # img = img0.as_numpy()
                img = letterbox(img, 640, stride=32)[0]

                # Convert
                img = img[:, :, ::-1].transpose(2, 0, 1)  # BGR to RGB, to 3x416x416
                img = np.ascontiguousarray(img)
                # img = torch.from_numpy(img)#.to(device)
                # img = img.half() if half else img.float()  # uint8 to fp16/32
                # img = img.float()
                img = img.astype(np.float32)
                img /= 255.0  # 0 - 255 to 0.0 - 1.0
                res.append(img)
            img = np.array(res)
            out_tensor_0 = pb_utils.Tensor("canopycoverage_preprocessing_output",
                                           img)

            inference_response = pb_utils.InferenceResponse(
                output_tensors=[out_tensor_0])
            responses.append(inference_response)
        print(f'preprocess time: {time.time() - t0}')
        return responses

    def finalize(self):
        """`finalize` is called only once when the model is being unloaded.
        Implementing `finalize` function is OPTIONAL. This function allows
        the model to perform any necessary clean ups before exit.
        """
        print('Cleaning up...')
```
BTW, I implemented batching with a for loop in excute, but that's not the problem, because currently the times printed are all batch 1.

```text
name: "canopycoverage_preprocess"
backend: "python"
max_batch_size: 32 
input [
{
    name: "canopycoverage_preprocessing_input"
    data_type: TYPE_UINT8
    dims: [ -1, -1, 3 ]
}
]
 
output [
{
    name: "canopycoverage_preprocessing_output"
    data_type: TYPE_FP32
    dims: [ 3, -1, -1 ]
}
]

instance_group [{ kind: KIND_CPU }]
```

Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).
python backend

**Expected behavior**
Normally, the pre-processing model takes about 1s to process a request.

**Comment by WangYizhang01** [comment]
I tried the examples from the tutorials(https://github.com/triton-inference-server/tutorials/tree/main/Conceptual_Guide/Part_5-Model_Ensembles) and it seems that the pre and post processing models do not write batches in excute func.I tried to send a request to it, and the request is a batch of pictures, and it seems to only process the first picture. So how does python backend support batch processing?

**Comment by manhtd98** [comment]
I am running batch with python backend in prod, the batch has not perfectly worked, some case it did not decrease to max batch size, but in most of case it work correctly.

**Comment by dyastremsky** [comment]
@manhtd98, that sounds like a different case. Triton should return an error if you try to send a request larger than the max_batch_size. If it's smaller than the max batch size, it may have to do with timing and you can try setting a [larger max_queue_delay](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_configuration.md#delayed-batching). If you're seeing different behavior, please open a separate issue.

@WangYizhang01, can you provide your client code as well? I used perf analyzer even with concurrency 1 and found no issues: 

`perf_analyzer -m canopycoverage_preprocess --shape canopycoverage_preprocessing_inpu
t:4,4,3 --shape canopycoverage_preprocessing_output:3,4,4
*** Measurement Settings ***
  Batch size: 1
  Service Kind: Triton
  Using "time_windows" mode for stabilization
  Measurement window: 5000 msec
  Using synchronous calls for inference
  Stabilizing using average latency

Request concurrency: 1
  Client: 
    Request count: 922
    Throughput: 51.2128 infer/sec
    Avg latency: 19516 usec (standard deviation 2979 usec)
    p50 latency: 20225 usec
    p90 latency: 21022 usec
    p95 latency: 21238 usec
    p99 latency: 22239 usec
    Avg HTTP time: 19499 usec (send/recv 6315 usec + response wait 13184 usec)
  Server: 
    Inference count: 922
    Execution count: 922
    Successful request count: 922
    Avg request latency: 12691 usec (overhead 17 usec + queue 90 usec + compute input 43 usec + compute infer 11243 usec + compute output 1298 usec)

Inferences/Second vs. Client Average Batch Latency
Concurrency: 1, throughput: 51.2128 infer/sec, latency 19516 usec`

**Comment by dyastremsky** [comment]
Closing due to inactivity. If you would like to reopen this issue for follow-up, please let us know.
